### ISSUE TITLE
Rated sessions do not display as such when revisited.

### ISSUE THREAD

**Issue by tomconnell-wf** [GoogleChrome/ioweb2016#1006]
I suspect that this starts happening when I get a new service worker.

Steps:
1. Go to a session page.
1. Rate the session.
1. Notice that the Rate Session button is grayed out, and cannot be clicked.  I believe the text changes in the button, also.
1. Close that chrome tab.
1. Return to that session in a new tab, the rate session button  is clickable again.
1. If you attempt to submit new feedback, you get a error message.

I did this on a Nexus 6p, mobile web.

**Comment by ebidel** [comment]
We're aware of this issue. For some reason the session ratings aren't being stored in FB.

@pengying

**Comment by pengying** [comment]
@crhym3 @nicolasgarnier 

We debugged it and it seems like the backend isn't updating firebase.

**Comment by crhym3** [comment]
The backend *is* updating firebase. otherwise, you wouldn't be able to complete steps 1-3.

**Comment by pengying** [comment]
Only occurs in prod :/.

**Comment by pengying** [comment]
Sorry for being short at the after party. I meant it works fine in stage. But is only broken in prod.

**Comment by crhym3** [comment]
So, in other words firebase security rules don't match. Is that what you're saying?

**Comment by ebidel** [comment]
Alex we need help debugging. Can you take a look?

**Comment by crhym3** [comment]
There's not much I see from the backend, just this:

```
401 Unauthorized
{"error" : "Permission denied"}
```

**Comment by pengying** [comment]
@nicolasgarnier can you take a look at the permissions?

**Comment by nicolasgarnier** [comment]
The permissions are the same on prod/dev/stage, I deployed the same rules automatically on stage/prod/dev typically and we deploy the same file.

You'll get a permission denied if the session has already marked as reviewed and you try to write to it though, as designed.

@crhym3 what were you trying to do when you got the error? Read or Write? Or did you just see this from the logs?
I'm guessing you are seeing this error from the logs, it must be coming from a tentative write on an already submitted session which is the intended behavior. That's how we check for already submitted sessions.

**Comment by nicolasgarnier** [comment]
I can find users with feedback submitted session in prod. for instance:

https://io-2016-prod-1.firebaseio.com/data/google:116640020808078754263

![image](https://cloud.githubusercontent.com/assets/3766663/15457574/04a4ca52-208c-11e6-9467-50d4afea927c.png)

**Comment by nicolasgarnier** [comment]
For now it seems that the data is indeed being written by the backend correctly. Peng you can check if the data is there by using the admin console for the user you are testing with:

https://io-2016-prod-<SHARD_NUM>.firebaseio.com/data/<UID>

If the data is there then this is likely just a UI issue.

**Comment by pengying** [comment]
Can you share the Firebase with me or check
https://io-2016-prod-4.firebaseio.com/data/google:115132759088093329854

app.savedSurveys is always empty for me.
IOWA.Elements.Template.app.savedSurveys
[]

**Comment by ebidel** [comment]
@pengying @nicolasgarnier , nothing is being saved:

<img width="349" alt="screen shot 2016-05-23 at 4 50 57 pm" src="https://cloud.githubusercontent.com/assets/238208/15488021/ca8af9c4-2106-11e6-8220-b59601eadf00.png">

**Comment by nicolasgarnier** [comment]
@pengying I shared all the prod DB with you

**Comment by nicolasgarnier** [comment]
I think i found what the issue is:

https://io-2016-prod-9.firebaseio.com/data/google:115132759088093329854

![image](https://cloud.githubusercontent.com/assets/3766663/15503212/e7c7be02-21b8-11e6-8911-342283c64705.png)

Basically it seems the Backend is not writing to the same shard as the frontend. Something must be off in the Shard calculation the backend does. I could check the code if you point me to it. Just as a wild guess are you guys appending `google:` to the user ID?

**Comment by nicolasgarnier** [comment]
I checked the Go backend code that finds the shard and it seems correct.

@pengying can you double check that you are passing the UID with no extra trailing/leading characters? Basically could you print out the HTTP request done to the server?

**Comment by nicolasgarnier** [comment]
Seems like the UID should always be passed correctly:
https://github.com/GoogleChrome/ioweb2016/blob/master/app/scripts/helper/schedule.js#L308-L312

**Comment by nicolasgarnier** [comment]
I think I got it:

The issue is actually with the Go code and specifically at this line:

https://github.com/GoogleChrome/ioweb2016/blob/master/backend/survey.go#L68

`gid := strings.TrimPrefix("google:", uid)`

should instead be:

`gid := strings.TrimPrefix(uid, "google:")`

You can try this here: https://play.golang.org/p/DD4ZDSHuZO

`gid := strings.TrimPrefix("google:", "google:115132759088093329854")`
prints: `google:` instead of `115132759088093329854`.

Basically right now all survey feedback is being written to shard 9...

**Comment by nicolasgarnier** [comment]
we coudl fix it on the backend but I think it;s basically a bit late to do that :D
Instead we could also just have the client look at shard 9 always for all session feedback data since it;s all there. i don't think the load is going to be an issue now that IO has passed.

**Comment by ebidel** [comment]
Is the client always going to have access to shard 9?

**Comment by nicolasgarnier** [comment]
yes we only restrict based on UID in the path which in this case match...

**Comment by ebidel** [comment]
This isn't huge priority now that IO is over, but that seems like the most logical thing to do atm now that the data is all in that location :\

@pengying if you want to fix this, I'm supportive :)

**Comment by pengying** [comment]
Haha.  I mean I think we'll reuse the infrastructure for next year right?  I think we should fix the backend and push out a version of the front end that looks at shard 9.  But as part of the repo we should probably keep the non jacked up versions.

**Comment by ebidel** [comment]
I like that plan
